src/serialization.jl

```diff
@@ -1,19 +1,63 @@
 # This is the code that hands the serialiation and deserialization of each object
 
+const FORMATTERS = (
+    # format = (
+    #    deserialize! # IO -> value
+    #    serialize! # IO, Value -> nothing
+    #)
+    bson = (
+        deserialize! = first ∘ values ∘ BSON.load,
+        serialize! = (io, value) -> bson(io, Dict("object" => value))
+    ),
+    julia_serialize = (
+        deserialize! = Serialization.deserialize,
+        serialize! = Serialization.serialize,
+    )
+)
+
+function formatter(format)
+    return get(FORMATTERS, format) do
+        error(LOGGER, ArgumentError("Unsupported format $(format)"))
+    end
+end
+
+
+const COMPRESSORS = (
+    none = (
+        compress = identity,
+        decompress = identity
+    ),
+    gzip = (
+        compress = GzipCompressorStream,
+        decompress = GzipDecompressorStream,
+    ),
+    gzip_fastest = (
+        compress = io -> GzipCompressorStream(io; level=1),
+        decompress = GzipDecompressorStream,
+    ),
+    gzip_smallest = (
+        compress = io -> GzipCompressorStream(io; level=9),
+        decompress = GzipDecompressorStream,
+    ),
+)
+
+function compressor(compression)
+    return get(COMPRESSORS, compression) do
+        error(LOGGER, ArgumentError("Unsupported compression $(compression)"))
+    end
+end
+
+
 """
     getindex(jlso, name)
 
 Returns the deserialized object with the specified name.
 """
 function Base.getindex(jlso::JLSOFile, name::String)
     try
-        if jlso.format === :bson
-            BSON.load(IOBuffer(jlso.objects[name]))[name]
-        elseif jlso.format === :serialize
-            deserialize(IOBuffer(jlso.objects[name]))
-        else
-            error(LOGGER, ArgumentError("Unsupported format $(jlso.format)"))
-        end
+        buffer = IOBuffer(jlso.objects[name])
+        decompressing_buffer = compressor(jlso.compression).decompress(buffer)
```

Comment:
I'm not sure I like the named tuple pattern either as it makes this code harder to read. I think maybe having `Formatter{:method}` and `Compressor{:method}` might be a nicer way of writing this?

```
struct Formatter{M}
    function Formatter{M}()
        M in (:bson, :julia_serialize) || error(LOGGER, ArgumentError("Unsupported format $M"))
    end
end

serialize!(f::Formatter{:bson}, io, value) = bson(io, Dict("object" => value))
deserialize!(f::Formatter{:bson}, io) = first(values(BSON.load(io)))
# Same methods for Formatter{:julia_serialize}
...

struct Compressor{M}
...
end

compress(c::Compressor{:gzip}, io) = GzipCompressorStream(io)
decompress(c::Compressor{:gzip}, io) = GzipDecompressorStream(io)
```

Then the client code can look more reasonable with:

```
buffer = IOBuffer(jlso.objects[name])
compressor = Compressor{jlso.compression}()
formatter = Formatter{jlso.format}()

return deserialize!(formatter, decompress(compressor, buffer))
```

Reply:
Also, I'm not sure I follow the pattern for using `!` here.

Reply:
We are mutating the stream.
All stream functions should have `!` but do not for historical reasons.

I like your proposal, the NamedTuple is basically a hack to allow for single dispatch on `Symbol`. But a type is much better.

I think it would be better if we didn't bother with the `ArgumentError`
since the method error that would result would convey all the same info -- including the alternatives.

Reply:
> All stream functions should have `!` but do not for historical reasons.

Ahh, okay, this is the part that always confuses me. I remember some folks suggesting that it depends on whether a function is mutating at the OS level (e.g., unix sockets)?

> I think it would be better if we didn't bother with the ArgumentError
since the method error that would result would convey all the same info -- including the alternatives.

👍

Reply:
I should probably not have them. That way I can be consistant with `Base`.